shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -0,0 +1,303 @@
+      try {
+        new ColumnarBatch(cols.map(_.copyToHost()).toArray, batch.numRows())
+      } finally {
+        cols.map(_.close())
+      }
+    })
+    cbRdd.mapPartitions(iter => new CloseableColumnBatchIterator(iter))
+  }
+
+  /**
+   * Convert the cached batch into `InternalRow`s.
+   * @param input the cached batches that should be converted.
+   * @param cacheAttributes the attributes of the data in the batch.
+   * @param selectedAttributes the field that should be loaded from the data and the order they
+   *                           should appear in the output rows.
+   * @param conf the configuration for the job.
+   * @return RDD of the rows that were stored in the cached batches.
+   */
+  override def convertCachedBatchToInternalRow(input: RDD[CachedBatch],
+                                cacheAttributes: Seq[Attribute],
+                                selectedAttributes: Seq[Attribute],
+                                conf: SQLConf): RDD[InternalRow] = {
+    val requestedColumnIndices = selectedAttributes.map(a =>
+      cacheAttributes.map(_.exprId).indexOf(a.exprId))
+    // if plugin enabled
+    val cbRdd: RDD[InternalRow] = input.map(batch => {
+      if (batch.isInstanceOf[ParquetCachedBatch]) {
+        val parquetCB = batch.asInstanceOf[ParquetCachedBatch]
+        withResource(Table.readParquet(ParquetOptions.DEFAULT, parquetCB.getBuffer(), 0,
+          parquetCB.sizeInBytes)) { table =>
+          withResource(GpuColumnVector.from(table)) { cb =>
+            val cols = GpuColumnVector.extractColumns(cb)
+            InternalRow(requestedColumnIndices.map(ordinal => cols(ordinal)).map(_.copyToHost()))
+          }
+        }
+      } else {
+        throw new IllegalStateException("I don't know how to convert this batch")
+      }
+    })
+    cbRdd
+  }
+
+  /**
+   * Convert an `RDD[InternalRow]` into an `RDD[CachedBatch]` in preparation for caching the data.
+   * We use the RowToColumnarIterator and convert each batch at a time
+   * @param input the input `RDD` to be converted.
+   * @param schema the schema of the data being stored.
+   * @param storageLevel where the data will be stored.
+   * @param conf the config for the query.
+   * @return The data converted into a format more suitable for caching.
+   */
+  override def convertInternalRowToCachedBatch(input: RDD[InternalRow],
+     schema: Seq[Attribute],
+     storageLevel: StorageLevel,
+     conf: SQLConf): RDD[CachedBatch] = {
+    val s = StructType(schema.map(a => StructField(a.name, a.dataType, a.nullable, a.metadata)))
+    val converters = new GpuRowToColumnConverter(s)
+    val numInputRows = SQLMetrics.createMetric(input.context, NUM_INPUT_ROWS)
+    val numOutputBatches = SQLMetrics.createMetric(input.context, NUM_OUTPUT_BATCHES)
+    val numOutputRows = SQLMetrics.createMetric(input.context, NUM_OUTPUT_ROWS)
+    val totalTime = SQLMetrics.createMetric(input.context, TOTAL_TIME)
```

Comment:
These metrics are not registered anywhere, so why are we creating them?  We should either find a way to register them so that the user can debug what is happening, or we should make `RowToColumanrIteraor` have them be optional.